File: src/correctpony.c
/**
 * correctpony - a passphrase generator inspired by http://xkcd.com/936/
 * Copyright (C) 2012, 2013  Mattias Andrée (maandree@member.fsf.org)
 * 
 * This program is free software. It comes without any warranty, to
 * the extent permitted by applicable law. You can redistribute it
 * and/or modify it under the terms of the Do What The Fuck You Want
 * To Public License, Version 2, as published by Sam Hocevar. See
 * http://sam.zoy.org/wtfpl/COPYING for more details.
 */
#include "correctpony.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>

#ifdef _GNU_SOURCE
#include <getopt.h>
#endif

#include "wlist.h"

static size_t rand_index(size_t n);
static void rand_perm(size_t* dest, size_t n);
static void print_usage(char* argv0);
static void print_version(char* argv0);


static size_t rand_index(size_t n)
{
  FILE* file = fopen("/dev/urandom", "r");
  
  int value = 0, i;
  for (i = 0; i < 4; i++)
    value = (value << 8) | fgetc(file);
  
  fclose(file);
  
  return value % n;
}

static void rand_perm(size_t* dest, size_t n)
{
  size_t i, j;
  int* check = malloc(n * sizeof *check);
  
  if (!check)
    return;
  
  memset(check, 0, n * sizeof *check);
  
  for (i = 0; i < n; ++i)
    {
      for (j = rand_index(n); check[j]; j = (j+1) % n)
	;
      
      dest[i] = j;
      check[j] = 1;
    }
  
  free(check);
}

static void print_usage(char* argv0)
{
  fprintf(stderr, "usage: %s [options] [count]\n", argv0);
  fprintf(stderr, "generates [count] passphrases, or 1 if no number specified\n");
  fprintf(stderr, "\n");
  
  #ifdef _GNU_SOURCE
    fprintf(stderr, "Available options: \n"
                    "-h, --help             print this text and exit\n"
                    "-v, --version          print version and exit\n"
	            "-p, --nocolour         do not colour the words in the passphrase\n"
                    "-j, --join             join all words\n"
                    "-i, --include <word>   include <word> in passphrase\n"
                    "-l, --list <list>      use wordlist <list>.\n");
    fprintf(stderr, "                       must be either absolute path or path relative to " WORDLIST_DIR ".\n"
                    "                       max. " STR(WLISTS_MAX) " wordlists can be specified;\n"
	            "                       if omitted, the default wordlist " WORDLIST_DIR "/" WORDLIST_DEFAULT " will be used.\n");
    fprintf(stderr, "-c, --char <n>         passphrase must be at least <n> characters long (default: " STR(CHARS_MIN_DEFAULT) ")\n"
                    "-w, --words <n>        passphrase must consist of at least <n> words (default: " STR(WORDS_MIN_DEFAULT) ")\n"
                    "-s, --sep <char>       use <char> as separator between words\n"
                    "-u, --camelcase        capitalize first letter of each word\n"
           );
  #else
    fprintf(stderr, "available options: \n"
                    "-h         print this text and exit\n"
                    "-v         print version and exit\n"
                    "-p         do not colour the words in the passphrase\n"
                    "-j         join all words\n"
                    "-i <word>  include <word> in passphrase\n"
                    "-l <list>  use wordlist <list>.\n"
                    "           must be either absolute path or path relative to " WORDLIST_DIR ".\n"
                    "           max. " STR(WLISTS_MAX) " wordlists can be specified;\n");
    fprintf(stderr, "           if omitted, the default wordlist " WORDLIST_DIR "/" WORDLIST_DEFAULT " will be used.\n"
                    "-c <n>     passphrase must be at least <n> characters long (default: " STR(CHARS_MIN_DEFAULT) ")\n"
                    "-w <n>     passphrase must consist of at least <n> words (default: " STR(WORDS_MIN_DEFAULT) ")\n"
                    "-s <char>  use <char> as separator between words\n"
                    "-u         capitalize first letter of each word\n"
           );
  #endif
}

static void print_version(char* argv0)
{
  printf("%s Version: " VERSION "\n", argv0);
  printf("Copyright (C) 2012, 2013  Mattias Andrée (maandree@member.fsf.org)\n");
  printf("License: WTFPL 2.0 http://sam.zoy.org/wtfpl/COPYING\n");
}

int main(int argc, char** argv)
{
  int c;
  size_t i;
  
  size_t n_phrases = 1;
  
  size_t out_idx[WORDS_MAX];
  
  /* words to choose from */
  char* words[WORDS_MAX];
  size_t n_words = 0, n_chars = 0;
  
  /* minimum lengths to generate */
  size_t chars_min = CHARS_MIN_DEFAULT;
  size_t words_min = WORDS_MIN_DEFAULT;
  
  /* user-specified wordlists */
  struct wlist* wlists[WLISTS_MAX];
  size_t i_wlists = 0, n_wlists = 0;
  
  /* user-specified words */
  char* uwords[UWORDS_MAX];
  size_t i_uwords = 0, n_uwords = 0;
  
  /* sep character between words, -1 -> none */
  int sep = 32; /* ' ' */
  int camel = 0;
  
  /* for strtoul */
  char* endptr;
  
  /* colour each word */
  int colour = 1;
  
  /* arguments */
  
  #ifdef _GNU_SOURCE
    struct option long_opts[] = {
        { "include",     required_argument, NULL, 'i' },
        { "list",        required_argument, NULL, 'l' },
        { "wordlist",    required_argument, NULL, 'l' },
        { "word-list",   required_argument, NULL, 'l' },
        { "chars",       required_argument, NULL, 'c' },
        { "words",       required_argument, NULL, 'w' },
        { "word",        required_argument, NULL, 'w' },
        { "sep",         required_argument, NULL, 's' },
        { "separator",   required_argument, NULL, 's' },
        { "uppercase",   no_argument, NULL, 'u' },
        { "camelcase",   no_argument, NULL, 'u' },
        { "upper-case",  no_argument, NULL, 'u' },
        { "camel-case",  no_argument, NULL, 'u' },
        { "join",        no_argument, NULL, 'j' },
        { "help",        no_argument, NULL, 'h' },
        { "version",     no_argument, NULL, 'v' },
        { "nocolour",    no_argument, NULL, 'p' },
        { "nocolor",     no_argument, NULL, 'p' },
        { "no-colour",   no_argument, NULL, 'p' },
        { "no-color",    no_argument, NULL, 'p' },
        { "nocolours",   no_argument, NULL, 'p' },
        { "nocolors",    no_argument, NULL, 'p' },
        { "no-colours",  no_argument, NULL, 'p' },
        { "no-colors",   no_argument, NULL, 'p' },
        { "pipe",        no_argument, NULL, 'p' },
    };
    while ((c = getopt_long(argc, argv, "hvjpi:l:c:w:s:u", long_opts, NULL)) != -1)
  #else
    while ((c = getopt(argc, argv, "hvjpi:l:c:w:s:u")) != -1)
  #endif
    {
      switch (c)
        {
	case 'h':
	  print_usage(argv[0]);
	  exit(EXIT_SUCCESS);
	  
	case 'v':
	  print_version(argv[0]);
	  exit(EXIT_SUCCESS);
	  
	case 'j':
	  sep = -1;
	  break;
	  
	case 'p':
	  colour = 0;
	  break;
	  
	case 'i':
	  uwords[i_uwords] = optarg;
	  n_uwords = (n_uwords < UWORDS_MAX) ? n_uwords + 1 : UWORDS_MAX;
	  i_uwords = (i_uwords + 1) % UWORDS_MAX;
	  break;
	  
	case 'l':
	  wlists[i_wlists] = wlist_read(optarg);
	  if (wlists[i_wlists])
	    {
	      n_wlists = (n_wlists < WLISTS_MAX) ? n_wlists + 1 : WLISTS_MAX;
	      i_wlists = (i_wlists + 1) % WLISTS_MAX;
	    }
	  break;
	  
	case 'c':
	  chars_min = strtoul(optarg, &endptr, 10);
	  if (*endptr != '\0')
	    {
	      fprintf(stderr, "invalid number: %s\n", optarg);
	      exit(EXIT_FAILURE);
	    }
	  chars_min = (chars_min > CHARS_MAX) ? CHARS_MAX : chars_min;
	  break;
	  
	case 'w':
	  words_min = strtoul(optarg, &endptr, 10);
	  if (*endptr != '\0')
	    {
	      fprintf(stderr, "invalid number: %s\n", optarg);
	      exit(EXIT_FAILURE);
	    }
	  words_min = (words_min > WORDS_MAX) ? WORDS_MAX : words_min;
	  break;
	  
	case 's':
	  sep = *optarg;
	  break;
	  
	case 'u':
	  camel = 1;
	  break;
	  
	case '?':
	  fprintf(stderr, "uknown option \"-%c\"\n", optopt);
	  exit(EXIT_FAILURE);
        }
    }
    
  if (optind < argc)
    {
      n_phrases = strtoul(argv[optind], &endptr, 10);
      if (*endptr != '\0')
	{
          fprintf(stderr, "invalid number: %s\n", argv[optind]);
	  exit(EXIT_FAILURE);
        }
    }
    
  /* no user-specified wordlists, use default */
  if (!n_wlists)
    {
      wlists[0] = wlist_read(WORDLIST_DEFAULT);
      if (!wlists[0])
	{
          fprintf(stderr, "couldn't read default wordlist\n");
          exit(EXIT_FAILURE);
	}
      n_wlists = 1;
    }
  
  while (n_phrases--)
    {
      n_chars = 0;
      n_words = 0;
      
      for (; n_words < n_uwords; ++n_words)
        {
          words[n_words] = uwords[n_words];
          n_chars += strlen(words[n_words]);
        }
      
      for (; n_words < words_min || n_chars < chars_min; ++n_words)
        {
          size_t iwl, iw;
          iwl = rand_index(n_wlists);
          
          iw = rand_index(wlists[iwl]->len);
          
          words[n_words] = wlist_get(wlists[iwl], iw);
          n_chars += strlen(words[n_words]);
        }
      
      if (camel)
        {
          for (i = 0; i < n_words; ++i)
            *words[i] = toupper(*words[i]);
        }
      
      rand_perm(out_idx, n_words);
      
      if (colour)
	printf("\033[0;1m");
      
      for (i = 0; i < n_words; ++i)
        {
          if (i && sep != -1)
	    printf("%c", sep);
	  
	  if (!colour)
	    printf("%s", words[out_idx[i]]);
	  else if (i & 1)
	    printf("\033[34m%s", words[out_idx[i]]);
	  else
	    printf("\033[33m%s", words[out_idx[i]]);
        }
      
      if (colour)
	printf("\033[0m\n");
      else
	printf("\n");
  }
  
  for (i = 0; i < n_wlists; ++i)
    wlist_free(wlists[i]);
  
  return EXIT_SUCCESS;
}

File: src/Correctpony.java
/**
 * correctpony — a passphrase generator inspired by xkcd 936
 * 
 * Copyright © 2012, 2013  Mattias Andrée (maandree@member.fsf.org)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import argparser.ArgParser;

import java.io.*;
import java.util.*;


/**
 * Mane class for correctpony
 * 
 * Mattias Andrée, <a href="mailto:maandree@member.fsf.org">maandree@member.fsf.org</a>
 */
public class Correctpony
{
    /**
     * ‘:’ separated list of directionary directories
     */
    public static final String DICT_DIRS = "/usr/share/dict:/usr/local/share/dict:~/share/dict";
    /* The makefile changes this line [DICT_DIRS], do not edit the value or name.  */
    
    /**
     * The version of the program
     */
    public static final String VERSION = "2.0";
    
    /**
     * Default word separators
     */
    public static final String[] DEFAULT_SEPARATORS = { " " };
    
    /**
     * Default dictionaries
     */
    public static final String[] DEFAULT_DICTIONARIES = null;
    
    
    
    /**
     * Mane method
     * 
     * @param  args  Command line arguments, excluding exec
     * 
     * @throws  IOException  On I/O error
     */
    public static void main(String... args) throws IOException
    {
	String usage = "[-p] [-r DEVICE] [-j...] [-s SEP...] [-u] [-c COUNT] [-w COUNT] [-i WORD...] [-l LIST...] [COUNT | -z]\n"
	             + "-h | --copying | --warranty | --version | --wordlists [--full] | --wordcount [-l LIST...]";
	for (final String symbol : new String[] { "[", "]", "(", ")", "|" })
	    usage = usage.replace(symbol, "\033[02m" + symbol + "\033[22m");
	usage = "\033[34mcorrectpony\033[00m " + usage.replace("\n", "\n\033[34mcorrectpony\033[00m ");
	ArgParser parser = new ArgParser("passphrase generator inspired by xkcd 936", usage,
					 "correctpony generates a phrase with real words, making it easy to remember\n" + 
					 "but hard to brute force. But the strenghten your security you may want to\n" +
					 "do an entroy check on the phrase, it should be enought to just read the phrase\n" +
					 "and make sure no close pairs is a common combination in read text. And if you\n" +
					 "are paranoid do one or two simple modifications to the generated phrase.");
	
	parser.add(new ArgParser.Argumentless(           1, "-h", "--help"),       "Print this help information");
	parser.add(new ArgParser.Argumentless(           0, "--copying"),          "Print copyright information");
	parser.add(new ArgParser.Argumentless(           0, "--warranty"),         "Print warranty information");
	parser.add(new ArgParser.Argumentless(           0, "--version"),          "Print the program's name and version");
	parser.add(new ArgParser.Argumentless(           0, "--wordlists"),        "Print all available wordlists");
	parser.add(new ArgParser.Argumentless(           0, "--full"),             "Print full file path");
	parser.add(new ArgParser.Argumentless(           0, "--wordcount"),        "Print the number of unique words available");
	parser.add(new ArgParser.Argumentless(           1, "-p", "--nocolour"),   "Do not print with colours");
	parser.add(new ArgParser.Argumented("DEVICE",    1, "-r", "--random"),     "Random number generator to use");
	parser.add(new ArgParser.Argumentless(           1, "-j", "--join"),       "Add word joining as a separator");
	parser.add(new ArgParser.Argumented("SEPARATOR", 1, "-s", "--separator"),  "Add a separator");
	parser.add(new ArgParser.Argumentless(           1, "-u", "--camelcase"),  "Capitalise first letter of each word");
	parser.add(new ArgParser.Argumented("COUNT",     1, "-c", "--characters"), "Least number of characters");
	parser.add(new ArgParser.Argumented("COUNT",     1, "-w", "--words"),      "Least number of words");
	parser.add(new ArgParser.Argumented("WORD",      1, "-i", "--include"),    "Word that must be included");
	parser.add(new ArgParser.Argumented("WORDLIST",  1, "-l", "--list"),       "Word list to use");
	parser.add(new ArgParser.Argumentless(           1, "-z", "--security"),   "Show stats and security estimates");
	
	parser.parse(args);
	final Map<String, String[]> opts = parser.opts;
	
	parser.testFiles(0, opts.get("--security") != null ? 0 : 1, 1);
	
	
	if (opts.get("--help") != null)
	{   parser.help();
	    return;
	}
	if (opts.get("--version") != null)
	{   System.out.println("correctpony " + VERSION);
	    return;
	}
	if (opts.get("--copying") != null)
	{
	    System.out.println("correctpony — a passphrase generator inspired by xkcd 936\n");
	    System.out.println("\n");
	    System.out.println("Copyright © 2012, 2013  Mattias Andrée (maandree@member.fsf.org)\n");
	    System.out.println("\n");
	    System.out.println("This program is free software: you can redistribute it and/or modify\n");
	    System.out.println("it under the terms of the GNU Affero General Public License as published by\n");
	    System.out.println("the Free Software Foundation, either version 3 of the License, or\n");
	    System.out.println("(at your option) any later version.\n");
	    System.out.println("\n");
	    System.out.println("This program is distributed in the hope that it will be useful,\n");
	    System.out.println("but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
	    System.out.println("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
	    System.out.println("GNU Affero General Public License for more details.\n");
	    System.out.println("\n");
	    System.out.println("You should have received a copy of the GNU Affero General Public License\n");
	    System.out.println("along with this program.  If not, see <http://www.gnu.org/licenses/>.\n");
	    return;
	}
	if (opts.get("--warranty") != null)
	{
	    System.out.println("This program is distributed in the hope that it will be useful,\n");
	    System.out.println("but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
	    System.out.println("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
	    System.out.println("GNU Affero General Public License for more details.\n");
	    return;
	}
	if (opts.get("--wordlists") != null)
	{
	    final boolean full = opts.get("--full") != null;
	    for (final String dictionary : getDictionaries())
		if (full)
		    System.out.println(dictionary);
		else
		    System.out.println(dictionary.substring(dictionary.lastIndexOf('/') + 1));
	    return;
	}
	if (opts.get("--wordcount") != null)
	{
	    String[] wordlists = new String[0];
	    if (opts.get("--list") != null)
		for (final String list : opts.get("--list"))
		{   System.arraycopy(wordlists, 0, wordlists = new String[wordlists.length + 1], 0, wordlists.length - 1);
		    wordlists[wordlists.length - 1] = list;
		}
	    String[] dictionaries = getDictionaries();
	    if (wordlists.length != 0)
		dictionaries = getFiles(wordlists, dictionaries);
	    final String[] dictionary = getWords(dictionaries);
	    System.out.println(dictionary.length);
	    return;
	}
	
	
	final boolean colours = opts.get("--nocolour") == null;
	final boolean camelcase = opts.get("--camelcase") != null;
	String[] separators = new String[0];
	String[] words = new String[0];
	String[] wordlists = new String[0];
	int minChars = opts.get("--characters") != null ? Integer.parseInt(opts.get("--characters")[0]) : 12;
	int minWords = opts.get("--words") != null ? Integer.parseInt(opts.get("--words")[0]) : 4;
	int count = parser.files.size() == 0 ? 1 : Integer.parseInt(parser.files.get(0));
	String randomgen = opts.get("--random") != null ? opts.get("--random")[0] : "/dev/urandom";
	
	if (opts.get("--join") != null)
	    for (final String _ : opts.get("--join"))
	    {	System.arraycopy(separators, 0, separators = new String[separators.length + 1], 0, separators.length - 1);
		separators[separators.length - 1] = "";
	    }
	
	if (opts.get("--separator") != null)
	    for (final String separator : opts.get("--separator"))
	    {	System.arraycopy(separators, 0, separators = new String[separators.length + 1], 0, separators.length - 1);
		separators[separators.length - 1] = separator;
	    }
	
	if (opts.get("--include") != null)
	    for (final String word : opts.get("--include"))
	    {	System.arraycopy(words, 0, words = new String[words.length + 1], 0, words.length - 1);
		words[words.length - 1] = word;
	    }
	
	if (opts.get("--list") != null)
	    for (final String list : opts.get("--list"))
	    {	System.arraycopy(wordlists, 0, wordlists = new String[wordlists.length + 1], 0, wordlists.length - 1);
		wordlists[wordlists.length - 1] = list;
	    }
	
	if (separators.length == 0)
	    separators = DEFAULT_SEPARATORS;
	if (wordlists.length == 0)
	    wordlists = null;
	
	
	if (opts.get("--security") != null)
	{
	    String[] dictionaries = getDictionaries();
	    if (wordlists != null)
		dictionaries = getFiles(wordlists, dictionaries);
	    final String[] dictionary = getWords(dictionaries);
	    final int wordcount = dictionary.length;
	    final int procent = wordcount / 5000;
	    
	    System.out.println();
	    System.out.println("You have a total of " + wordcount + " words in all your selected word lists,");
	    System.out.println("The English language contains about 500000 words. That is c:a " + procent + " %.");
	    System.out.println();
	    
	    double time = (double)wordcount;
	    int minwords = Math.max(0, minWords - words.length);
	    time = Math.pow(time, minwords);
	    for (int i = 1; i <= words.length; i++)
		time *= minwords + i;
	    time *= Math.pow(separators.length, minwords + words.length - 1);
	    time /= 60 * 60 * 24 * 365.2425;
	    
	    System.out.println("Trying to crack the passphrase knowing all settings and the dictionaries,");
	    System.out.println("it would take about " + time / 1000000 + " years at most, making 1 000 000 guesses per seconds.");
	    System.out.println();
	    
	    System.out.println("Trying to crack the passphrase knowing all settings and the dictionaries,");
	    System.out.println("it would take about " + time / 1000 + " years at most, making 1000 guesses per seconds.");
	    System.out.println();
	    
	    System.out.println("Trying to crack the passphrase knowing all settings and the dictionaries,");
	    System.out.println("it would take about " + time * 5 + " years at most, making 1 guess every fifth second.");
	    System.out.println();
	    
	    return;
	}
	
	Correctpony.randomgen = new FileInputStream(randomgen);
	try
	{   final String[] passphrases = generatePassphrases(colours, camelcase, separators, words, wordlists, minChars, minWords, count);
	    for (final String passphrase : passphrases)
		System.out.println(passphrase);
	}
	finally
	{    Correctpony.randomgen.close();
	}
    }
    
    
    
    /**
     * Random number generatore character device input stream
     */
    public static InputStream randomgen = null;
    
    
    
    /**
     * Gets all dictionary files
     * 
     * @return  All dictionary files
     * 
     * @throws  IOException  On I/O error
     */
    public static String[] getDictionaries() throws IOException
    {
	String[] rc = new String[128];
	int rcptr = 0;
	
	String HOME = System.getenv("HOME");
	if ((HOME == null) || (HOME.length() == 0))
	    HOME = System.getProperty("user.home");
	
	for (String dir : DICT_DIRS.split(":"))
	{
	    if (dir.startsWith("~"))
		dir = HOME + dir.substring(1);
	    File fdir = new File(dir);
	    if (fdir.exists())
		fdir = fdir.getCanonicalFile();
	    if (fdir.exists() && fdir.isDirectory())
		for (String file : fdir.list())
		{
		    if (rcptr == rc.length)
			System.arraycopy(rc, 0, rc = new String[rc.length << 1], 0, rcptr);
		    file = (dir + "/" + file).replace("//", "/");
		    if ((new File(file)).isFile())
			rc[rcptr++] = file;
		}
	}
	
	System.arraycopy(rc, 0, rc = new String[rcptr], 0, rcptr);
	return rc;
    }
    
    
    /**
     * Gets dictionaries' filename by they name
     * 
     * @parma   dictionaries  The dictionaries by names
     * @param   files         All dictionaries by filename
     * @return                The found dictionaries by filename
     * 
     * @throws  IOException  On I/O error
     */
    public static String[] getFiles(String[] dictionaries, String[] files) throws IOException
    {
	final String[] candidates = new String[files.length];
	for (int i = 0, n = files.length; i < n; i++)
	    candidates[i] = "/" + files[i];
	
	String[] rc = new String[dictionaries.length];
	int ptr = 0;
	
	for (String dictionary : dictionaries)
	{   dictionary = "/" + dictionary;
	    for (final String candidate : candidates)
		if (candidate.endsWith(dictionary))
		{
		    if (ptr == rc.length)
			System.arraycopy(rc, 0, rc = new String[rc.length << 1], 0, ptr);
		    rc[ptr++] = candidate;
	}	}
	
	System.arraycopy(rc, 0, rc = new String[ptr], 0, ptr);
	return rc;
    }
    
    
    /**
     * Gets all unique words in any number of dictionaries
     * 
     * @param   dictionaries  Dictionaries by filename, {@code null} for all
     * @return                Unique words, in lower case
     * 
     * @throws  IOException  On I/O error
     */
    public static String[] getWords(String[] dictionaries) throws IOException
    {
	if (dictionaries == null)
	    dictionaries = getDictionaries();
	
	final HashSet<String> unique = new HashSet<String>();
	for (final String dictionary : dictionaries)
	    {
		final InputStream is = new FileInputStream(dictionary);
		byte[] buf;
		int ptr = 0;
		try
		{   buf = new byte[is.available()];
		    if (buf.length == 0)
			buf = new byte[1 << 13];
		    forever:
		        for (int n;;)
			{   while (ptr < buf.length)
			    {
				ptr += n = is.read(buf, ptr, buf.length - ptr);
				if (n <= 0)
				    break forever;
			    }
			    System.arraycopy(buf, 0, buf = new byte[buf.length << 1], 0, ptr);
		}	}
		finally
		{   try
		    {   is.close();
		    }
		    catch (final Throwable ignore)
		    {   /* ignore */
		}   }
		for (final String word : (new String(buf, 0, ptr, "UTF-8")).split("\n"))
		    if (word.length() != 0)
			unique.add(word/*.toLowerCase()*/);
	    }
	
	final String[] rc = new String[unique.size()];
	int ptr = 0;
	for (final String word : unique)
	     rc[ptr++] = word;
	return rc;
    }
    
    
    /**
     * Joins words
     * 
     * @param   words       The words to join
     * @param   separators  Separators
     * @param   colour      Whether to use colours
     * @return              The result
     * 
     * @throws  IOException  On I/O error
     */
    public static String join(String[] words, String[] separators, boolean colour) throws IOException
    {
	final StringBuilder rc = new StringBuilder();
	int i = 0;
	if (colour)
	    rc.append("\033[01m");
	for (final String word : words)
	{
	    if (i > 0)
	    {	if (colour)
		    rc.append("\033[31m");
		rc.append(separators[random(separators.length)]);
	    }
	    if (colour)
		rc.append((i & 1) == 0 ? "\033[34m" : "\033[33m");
	    rc.append(word);
	    i++;
	}
	if (colour)
	    rc.append("\033[00m");
	return rc.toString();
    }
    
    
    /**
     * Gets a random integer
     * 
     * @param   max  The exclusive upper bound
     * @return       A random non-negative integer
     * 
     * @throws  IOException  On I/O error
     */
    public static int random(int max) throws IOException
    {
	int rc = Correctpony.randomgen.read() & 127;
	rc = (rc << 8) | Correctpony.randomgen.read();
	rc = (rc << 8) | Correctpony.randomgen.read();
	rc = (rc << 8) | Correctpony.randomgen.read();
	if (rc == -1)
	{   System.err.println("End of random number generator file stream reached. USE A REAL RANDOM DEVICE!");
	    System.exit(2);
	}
	return rc % max;
    }
    
    
    /**
     * Generate passphrases
     * 
     * @param  colours     Include colours
     * @param  camelcase   Capitalise first letter of each word
     * @param  separators  Words separators
     * @param  words       Words that must be included
     * @param  wordlists   Word lists to use
     * @param  minChars    Minimum lenght
     * @param  minWords    Minimum number of words
     * @param  count       Number of passphrases to generate
     * 
     * @throws  IOException  On I/O error
     */
    public static String[] generatePassphrases(boolean colours, boolean camelcase, String[] separators, String[] words, String[] wordlists, int minChars, int minWords, int count) throws IOException
    {
	String[] dictionaries = getDictionaries();
	if (wordlists != null)
	    dictionaries = getFiles(wordlists, dictionaries);
	final String[] dictionary = getWords(dictionaries);
	
	int joinMinimum = (1 << 31) - 1;
	for (final String separator : separators)
	    if (joinMinimum > length(separator))
		joinMinimum = length(separator);
	
	final String[] passphrases = new String[count];
	for (int i = 0; i < count; i++)
	{
	    final ArrayList<String> passphrase = new ArrayList<String>();
	    int passphraseLength = -joinMinimum;
	    
	    minWords -= words.length;
	    for (final String word : words)
		minChars -= length(word);
	    minChars -= passphraseLength * words.length;
	    
	    while ((passphraseLength < minChars) || (passphrase.size() < minWords))
	    {	final String word = dictionary[random(dictionary.length)].toLowerCase();
		passphraseLength += length(word) + joinMinimum;
		passphrase.add(word);
	    }
	    
	    for (final String word : words)
	    {	int position = random(passphrase.size() + 1);
		passphrase.add(position, word.toLowerCase());
	    }
	    
	    final String[] passphrase_array = new String[passphrase.size()];
	    passphrase.toArray(passphrase_array);
	    passphrases[i] = join(passphrase_array, separators, colours);
	}
	return passphrases;
    }
    
    
    /**
     * Gets the actual length of string
     * 
     * @param   text  The string
     * @return        The length of the string
     */
    public static int length(String text)
    {
	/* This code is Java specific, Java encodes in UTF-16 */
	
	int rc = 0;
	for (int i = 0, n = text.length(); i < n; i++)
	{   char c = text.charAt(i);
	    if ((c & 0xFC00) != 0xDC00)
		rc++;
	}
	return rc;
    }
    
    
    /* TODO determine hyphenation for words to determine which letters may be uppercase */
    /* TODO occationally use letter translation */
    
    
}

